rustworkx-core/src/coloring.rs

```diff
@@ -295,6 +569,78 @@ where
     edge_colors
 }
 
+/// Color edges of a graph using a greedy graph coloring algorithm with preset
+/// colors.
+///
+/// This function works by greedily coloring the line graph of the given graph.
+///
+/// The coloring problem is NP-hard and this is a heuristic algorithm
+/// which may not return an optimal solution.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on.
+/// * `preset_color_fn` - A callback function that will receive the edge identifier
+///     for each edge in the graph and is expected to return an `Option<usize>`
+///     (wrapped in a `Result`) that is `None` if the edge has no preset and
+///     the usize represents the preset color.
+/// * `strategy` - The greedy strategy used by the algorithm.
+///
+/// # Example
+/// ```rust
+///
+/// use petgraph::graph::Graph;
+/// use petgraph::graph::EdgeIndex;
+/// use petgraph::Undirected;
+/// use rustworkx_core::dictmap::*;
+/// use rustworkx_core::coloring::{greedy_edge_color_with_coloring_strategy, ColoringStrategy};
+/// use std::convert::Infallible;
+///
+/// let g = Graph::<(), (), Undirected>::from_edges(&[(0, 1), (1, 2), (0, 2), (2, 3)]);
+/// let preset_color_fn = |_| Ok::<Option<usize>, Infallible>(None);
+/// let colors = greedy_edge_color_with_coloring_strategy(&g, preset_color_fn, ColoringStrategy::Degree).unwrap();
+/// let mut expected_colors = DictMap::new();
+/// expected_colors.insert(EdgeIndex::new(0), 2);
+/// expected_colors.insert(EdgeIndex::new(1), 0);
+/// expected_colors.insert(EdgeIndex::new(2), 1);
+/// expected_colors.insert(EdgeIndex::new(3), 2);
+/// assert_eq!(colors, expected_colors);
+/// ```
+///
+pub fn greedy_edge_color_with_coloring_strategy<G, F, E>(
+    graph: G,
+    preset_color_fn: F,
```

Comment:
Is there a reason to not make this `Option<F>` since it's not mandatory. It just looks a bit awkward in the `rustworkx` crate side  to handle the case when there isn't preset.

Reply:
Thanks! I definitely agree it's better to make it `Option<F>`. With this change, the function above would look like
```
pub fn greedy_edge_color_with_coloring_strategy<G, F, E>(
    graph: G,
    preset_color_fn: Option<F>,
    strategy: ColoringStrategy,
) -> Result<DictMap<G::EdgeId, usize>, E>
where
    G: EdgeCount + IntoNodeIdentifiers + IntoEdges,
    G::EdgeId: Hash + Eq,
    F: Fn(G::EdgeId) -> Result<Option<usize>, E>,
```
So far so good. I thought that using this function (from say testing) now becomes something like
```
let preset_color_fn = |node_idx: EdgeIndex| -> Result<Option<usize>, Infallible> {
    if node_idx.index() == 1 {
        Ok(Some(1))
    } else {
        Ok(None)
    }
};
let colors = greedy_edge_color_with_coloring_strategy(
        &graph,
        Some(preset_color_fn),
        ColoringStrategy::Degree,
    )
    .unwrap();
```
for the case that we want to specify the preset coloring function, and 
```
let colors = greedy_edge_color_with_coloring_strategy(
        &graph,
        None,
        ColoringStrategy::Degree,
    )
    .unwrap();
```
for the case that we do not. However the compiler does not like the second case:
```
    --> rustworkx-core\src\coloring.rs:1505:13
     |
1502 |         let colors = greedy_edge_color_with_coloring_strategy(
     |                      ---------------------------------------- required by a bound introduced by this call
...
1505 |             None,
     |             ^^^^ cannot infer type of the type parameter `T` declared on the enum `Option`
     |
     = note: multiple `impl`s satisfying `_: Fn<(petgraph::prelude::EdgeIndex,)>` found in the following crates: `alloc`, `core`:
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `greedy_edge_color_with_coloring_strategy`
help: consider specifying the generic argument
     |
1505 |             None::<T>,
     |                 +++++

```
and I can't figure out how to make this work.

Reply:
Ok, yeah I can see why rust isn't happy. It needs to know what the `Some()` branch is in that case so it can generate the code for the generic usage in this case. But `None` doesn't have the context and specifying the type with `None::<T>` would be super ugly, especially for a fn generic. I thought it would be more ergonomic to be able to do `None` here, but I guess if it's not that easy. Let's just leave this as is.